[STORM-1157] Adding dynamic profiling for worker, restarting worker, jstack, heap dump, and profiling

kishorvpatil opened the pull request:
- Request restarting worker.
- Request jstack, 
- Request heap dump, 
- Start Profiling
   - Dump profiling
   - Stop Profiling

revans2 commented:
It looks good to me.  I am a little bit concerned that there is no way to turn this off, partially or all the way.  So for windows where there is currently no support for this we will end up trying to run a shell script that will fail, and if someone does not have/want commercial java support from Oracle, i.e. using open JDK, or IBM JDK, etc.  They have to make code modifications, not just config modifications to not show users buttons that will not work.

Could we put in something that could let users turn off part or all of this feature?  At a minimum in the UI, but it would be good if nimbus could also let users know that this particular feature is not currently configured on, and if the worker could see the config and know to add in ```-XX:+UnlockCommercialFeatures -XX:+FlightRecorder``` on the command line.

Other then that I am +1.

kishorvpatil commented:
@revans2 Thank you for the feedback. Following changes should help make this feature completely optional and pluggable.   
   - Added `worker.profiler.enabled` to shutdown use of profilers - 
     - If disabled, UI simply removes the section **Profiling and Debugging** section. 
     - If disabled, the supervisor would not add "worker.profiler.childopts" to worker JVM options - hence avoiding issues with jvm that does not provide commercial JProfiler. 
   - Added `worker.profiler.command` to specify which script to use at supervisor level so that it can be customized for different profiler scheme.

kishorvpatil commented:
Currently, this feature lacks support for windows and command-line interface.

kishorvpatil commented:
<img width="1354" alt="dynamic_profiling_debugging_1" src="https://cloud.githubusercontent.com/assets/6090397/10902398/ec51a46a-81c2-11e5-92f5-d4c79d0b80df.png">
<img width="1366" alt="dynamic_profiling_debugging_2" src="https://cloud.githubusercontent.com/assets/6090397/10902399/ef7e868a-81c2-11e5-9f1d-ce85c33afa3e.png">
<img width="731" alt="dynamic_profiling_debugging_3" src="https://cloud.githubusercontent.com/assets/6090397/10902400/f18d7c6a-81c2-11e5-96a7-f5a7fc11a98f.png">

revans2 commented:
+1 it looks like a great addition.

zhuoliu commented on docs/DYNAMIC_WORKER_PROFILING.md:
```diff
@@ -0,0 +1,29 @@
+Dynamic Worker Profiling
```
Should we move this doc file to docs/documentation directory?
Also the DYNAMIC_LOG_LEVEL_SETTINGS.md file.

kishorvpatil commented:
@zhuoliu I moved docs under documentation location. Also, added pointers from main page.

zhuoliu commented on storm-core/src/clj/backtype/storm/ui/core.clj:
```diff
@@ -986,6 +1021,127 @@
         (.setLogConfig nimbus id new-log-config)
         (json-response (log-config id) (m "callback")))))
+
+  (GET "/api/v1/topology/:id/profiling/start/:host-port/:timeout"
```
We may need to update the REST API doc for this.

zhuoliu commented on storm-core/src/clj/backtype/storm/ui/core.clj:
```diff
@@ -986,6 +1021,127 @@
+                                       id)}
+                          (m "callback")))))
+
+  (GET "/api/v1/topology/:id/profiling/stop/:host-port"
```
Same here and below.

hustfxj replied:
I am concern about Thread concurrency. If user is setting the worker's ProfileAction, supervisor is reading the worker's ProfileAction. Then what happened？

kishorvpatil replied:
As rest of the communication between nimbus and supervisor for assignments - going through ZK answers all Thread concurrency issues. As user ProfileAction is written to ZK by nimbus ( which performs 2 phase commit). The supervisor can read old value or new depending on whether it reads is before or after the commit phase.. The supervisor reads it every 30 seconds.

zhuoliu commented:
Just some nits about docs. Otherwise looks really good to me.

kishorvpatil commented:
@zhuoliu I have updated the REST API documentation for all new endpoints.

zhuoliu commented:
+1

revans2 commented on STORM-UI-REST-API.md:
```diff
@@ -636,6 +636,201 @@ Sample response:
+
+|Parameter |Value   |Description  |
+|----------|--------|-------------|
+|id   	   |String (required)| Topology Id  |
```
Do we want an `:id` here to be consistent with the other Parameters?

revans2 commented on STORM-UI-REST-API.md:
```diff
@@ -636,6 +636,201 @@ Sample response:
+
+|Parameter |Value   |Description  |
+|----------|--------|-------------|
+|id   	   |String (required)| Topology Id  |
```
Here too.

revans2 commented on STORM-UI-REST-API.md:
```diff
@@ -636,6 +636,201 @@ Sample response:
+
+|Parameter |Value   |Description  |
+|----------|--------|-------------|
+|id   	   |String (required)| Topology Id  |
```
And here

revans2 commented:
The changes look good and I am +1 on merging this in.  I had one minor comment on the documentation, but it is not critical and I am happy to have this merged in, with or without the change.

kishorvpatil commented:
@revans2  Fixed REST API documentation for consistency. Moved it to right location and added necessary link for Documentation.md

revans2 commented:
Looks good +1
